### PR TITLE
Override DisposeAsync in PipeWriterStream

### PR DESCRIPTION
**Pull request by wtgodbe**
Addresses https://github.com/dotnet/runtime/issues/64326 to stop us doing sync over async

### CODE REVIEW

File: src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs [hunk 1]
```diff
@@ -27,6 +27,15 @@ protected override void Dispose(bool disposing)
...
         }
 
+        public async override ValueTask DisposeAsync()
+        {
+            if (!LeaveOpen)
+            {
+                await _pipeWriter.CompleteAsync().ConfigureAwait(false);
+            }
+            await base.DisposeAsync().ConfigureAwait(false);
+        }
```
Comment (halter73) [inline src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs:37]: Can we add a test case to PipeWriterStreamTests?

**Comment by wtgodbe** [comment]
Having trouble building locally, so pushing an un-tested test for CI to run while I figure that out

**Comment by halter73** [comment]
> In the issue:
> 
> > Also, the implementation of PipeWriterStream.Dispose assumes _pipeWriter.Complete(); doesn't throw. Is it a safe assumption and shouldn't it try/catch exceptions to ensure the stream resources are released no matter what?
> 
> Did we determine this is a non-issue?

PipeWriterStream extends System.IO.Stream directly and is `internal sealed` so there are never going to be other resources to be released. We arguably don't need to call `base.DisposeAsync()` at all though we might as well for consistency. It might be good hygiene to try/catch in general. I don't think it's something we need to start doing in this PR, but I wouldn't object to it.

**Comment by halter73** [comment]
I know DisposeAsync() shouldn't be on the hot path, but I am tempted to remove the `base.DisposeAsync()` (and `base.Dispose()` call in the other method while we're at it) and get rid of the async state machine.

**Comment by wtgodbe** [comment]
> I know DisposeAsync() shouldn't be on the hot path, but I am tempted to remove the base.DisposeAsync() (and base.Dispose() call in the other method while we're at it) and get rid of the async state machine.

Not sure I understand the motivation here, what alternative are you suggesting?

**Comment by BrennanConroy** [comment]
Basically, get rid of the explicit `async await` by removing the call to `base.DisposeAsync()`

**Review by halter73 — APPROVED** [review]

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**